tests/h/models/document/_meta_test.py

```diff
@@ -10,197 +11,95 @@
-            session=db_session,
-            claimant=claimant,
-            type=type_,
-            value=value,
-            document=document,
-            created=created,
-            updated=updated,
-        )
+    @pytest.mark.parametrize(
+        "doc_title,final_title",
+        ((None, "attr_title"), ("", "attr_title"), ("doc_title", "doc_title")),
+    )
+    def test_it_denormalizes_title_to_document_when_falsy(
+        self, db_session, meta_attrs, doc_title, final_title
+    ):
+        meta_attrs["value"] = ["attr_title"]
+        meta_attrs["document"] = document = Document(title=doc_title)
 
-        document = db_session.query(Document).get(document.id)
-        assert document.title == value[0]
-
-    def test_it_denormalizes_title_to_document_when_empty(self, db_session):
-        claimant = "http://example.com/claimant"
-        type_ = "title"
-        value = ["the title"]
-        document = Document(title="")
-        created = yesterday()
-        updated = datetime.now()
         db_session.add(document)
 
-        create_or_update_document_meta(
-            session=db_session,
-            claimant=claimant,
-            type=type_,
-            value=value,
-            document=document,
-            created=created,
-            updated=updated,
-        )
+        create_or_update_document_meta(session=db_session, **meta_attrs)
 
         document = db_session.query(Document).get(document.id)
-        assert document.title == value[0]
+        assert document.title == final_title
 
-    def test_it_skips_denormalizing_title_to_document_when_already_set(
-        self, db_session
+    def test_it_logs_a_warning_with_existing_meta_on_a_different_doc(
+        self, log, mock_db_session, factories, meta_attrs
     ):
-        claimant = "http://example.com/claimant"
-        type_ = "title"
-        value = ["the title"]
-        document = Document(title="foobar")
-        created = yesterday()
-        updated = datetime.now()
-        db_session.add(document)
-
-        create_or_update_document_meta(
-            session=db_session,
-            claimant=claimant,
-            type=type_,
-            value=value,
-            document=document,
-            created=created,
-            updated=updated,
-        )
-
-        document = db_session.query(Document).get(document.id)
-        assert document.title == "foobar"
-
-    def test_it_logs_a_warning(self, log, mock_db_session, factories):
-        """
-        It should warn on document mismatches.
-
-        It should warn if there's an existing DocumentMeta with a different
-        Document.
-        """
         document_one = factories.Document()
         document_two = factories.Document()
         existing_document_meta = factories.DocumentMeta(document=document_one)
-
         mock_db_session.query.return_value.filter.return_value.one_or_none.return_value = (
             existing_document_meta
         )
```

Comment:
I think I'll leave this one for now, purely because this was exhausting! But yes, a slightly less "I ran the code in exactly this way" style of testing is better.